File: CSX/Callbacks & Higher-order Functions/eitherCallback.js
//EITHERCALLBACK 

//expected input: callback1, callback2
//expected result: return a function that evaluates whether either callback is true
//expected output: callback

function eitherCallback(callback1, callback2) {
  function callbackCheck(n) {
    return (callback1(n) || callback2(n))
  }
  return callbackCheck;
}

function filterArray(array, callback) {
  const newArray = [];
  for (let i = 0; i < array.length; i += 1) {
    if (callback(array[i], i, array)) newArray.push(array[i]);
  }
  return newArray;
}
const arrOfNums = [10, 35, 105, 9];
const integerSquareRoot = n => Math.sqrt(n) % 1 === 0;
const over100 = n => n > 100;
const intSqRtOrOver100 = eitherCallback(integerSquareRoot, over100);
console.log(filterArray(arrOfNums, intSqRtOrOver100)); // should log: [105, 9]

File: CSX - Python/Callbacks & Higher-order Functions/either_callback.py
from math import sort

def filter_list(ls, cb):
    return [x for x in ls if cb(x)]

def either_callback(cb1, cb2):
    return lambda x: cb1(x) or cb2(x)

list_of_nums = [10, 35, 105, 9]

def integer_square_root(x):
    return sqrt(x) % 1 == 0

def over_100(x):
    return x > 100

int_sq_rt_or_over_100 = either_callback(integer_square_root, over_100)
print(filter_list(list_of_nums, int_sq_rt_or_over_100))

